### PR TITLE
PEP 589: TypedDict: Type Hints for Dictionaries with a Fixed Set of Keys

### PR DESCRIPTION
**Pull request by JukkaL**
First draft.

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately we couldn't find an account corresponding to your GitHub username on [bugs.python.org](https://bugs.python.org/) (b.p.o) to verify you have signed the CLA (this might be simply due to a missing "GitHub Name" entry in your b.p.o account settings). This is necessary for legal reasons before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.


You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for your contribution, we look forward to reviewing it!

File: pep-0589.rst [hunk 1]
```diff
@@ -0,0 +1,645 @@
...
+to allow experimentation with a wide variety of different type
+checking approaches.  In particular, type compatibility should be
+based on structural compatibility: a more specific TypedDict type can
+be compatible with a smaller (more general) TypedDict type.
+
+
+Class-based Syntax
+------------------
+
+A TypedDict type can be defined using the class defition syntax with
```
Comment (JelleZijlstra) [inline pep-0589.rst]: definition
```suggestion
A TypedDict type can be defined using the class definition syntax with
```

File: pep-0589.rst [hunk 2]
```diff
@@ -0,0 +1,645 @@
...
+    from typing import TypedDict
+
+    class Movie(TypedDict):
+        name: str
+        year: int
+
+``Movie`` is a TypedDict type with two items: ``'name'`` (with type
+``str``) and ``'year'`` (with type ``int``).
+
+A type checker should validate that the body a class-based TypedDict
```
Comment (JelleZijlstra) [inline pep-0589.rst]:
```suggestion
A type checker should validate that the body of a class-based TypedDict
```

File: pep-0589.rst [hunk 3]
```diff
@@ -0,0 +1,645 @@
...
+definition conforms to these rules:
+
+* The class body should only contain an optional docstring, followed
+  by lines with item definitions of form ``key: value_type``.  The
```
Comment (JelleZijlstra) [inline pep-0589.rst]:
```suggestion
  by lines with item definitions of the form ``key: value_type``.  The
```

File: pep-0589.rst [hunk 4]
```diff
@@ -0,0 +1,645 @@
...
+``isinstance()`` tests such as ``isinstance(d, Movie)``. The reason is
+that there is no existing support for checking types of dictionary
+item values, since ``isinstance()`` does not work with many PEP 484
+types, including common ones like ``List[str]``.  This would be needed
+for cases like this::
+
+    class Strings(TypedDict):
+        items: List[str]
+
+    print(isinstance({'items': [1]}))    # Should be False
```
Comment (JelleZijlstra) [inline pep-0589.rst]:
```suggestion
    print(isinstance({'items': [1]}, Strings))    # Should be False
```

File: pep-0589.rst [hunk 5]
```diff
@@ -0,0 +1,645 @@
...
+    print(isinstance({'items': ['x']}))  # Should be True
```
Comment (JelleZijlstra) [inline pep-0589.rst]:
```suggestion
    print(isinstance({'items': ['x']}, Strings))  # Should be True
```

File: pep-0589.rst [hunk 6]
```diff
@@ -0,0 +1,645 @@
...
+
+The above use case is not supported.  This is consistent with how
+``isinstance()`` is not supported for ``List[str]``.
+
+
+Inheritance
+-----------
+
+It is possible to inherit from one or more TypedDict types using the
+class-based syntax.  A single class cannot inherit from both a
```
Comment (JelleZijlstra) [inline pep-0589.rst]: The second sentence of the paragraph feels a bit out of place; it sounds like the example is going to be of inheritance from a TypedDict and a non-TypedDict.

File: pep-0589.rst [hunk 7]
```diff
@@ -0,0 +1,645 @@
...
+
+    Movie = TypedDict('Movie', {'name': str, 'year': int})
+
+It is also possible to specify totality using the alternative syntax::
+
+    Movie = TypedDict('Movie', {'name': str,
+                                'year': int}, total=False)
+
+The semantics are equivalent to the class-based syntax.  This syntax
+doesn't support inheritance, however, and there is no way to way to
```
Comment (JelleZijlstra) [inline pep-0589.rst]:
```suggestion
doesn't support inheritance, however, and there is no way to
```

File: pep-9999.rst [hunk 8]
```diff
@@ -0,0 +1,645 @@
+PEP: 589
```
Comment (JelleZijlstra) [inline pep-9999.rst]: Remove this copy of the file?

**Comment by brettcannon** [comment]
FYI @JukkaL you're flagged as having not signed the CLA and CI failed.

File: pep-0589.rst [hunk 9]
```diff
@@ -0,0 +1,648 @@
+PEP: 589
+Title: TypedDict: Type Hints for Dictionaries with a Fixed Set of Keys
+Author: Jukka Lehtosalo <jukka.lehtosalo@iki.fi>
+Sponsor: Guido van Rossum <guido@python.org>
+Discussions-To: typing-sig@python.org
```
Comment (ilevkivskyi) [inline pep-0589.rst]: It looks like this has not been posted to typing-sig yet. I think this looks very solid and is ready to be posted there.

**Pull request merged**